File: clients/ts/tests/common/harness/whirlpool.ts
import type { Address } from "@solana/addresses";

import { assertTailLength, venues, type VenueWindow } from "../../../src/index.js";
import { slotAt, type Fixture } from "../fixture.js";
import { key } from "../keys.js";
import { assertProgram } from "./index.js";

const { PROGRAM_ADDRESS, resolve: build } = venues.whirlpool;

const BASE_SLOT_COUNT = 15;

function placeholderAccounts(): venues.whirlpool.WhirlpoolAccounts {
  return {
    tokenProgramA: key(1),
    tokenProgramB: key(2),
    tokenAuthority: key(3),
    whirlpool: key(4),
    mintA: key(5),
    mintB: key(6),
    tokenOwnerAccountA: key(7),
    tokenVaultA: key(8),
    tokenOwnerAccountB: key(9),
    tokenVaultB: key(10),
    tickArray0: key(11),
    tickArray1: key(12),
    tickArray2: key(13),
    oracle: key(14),
  };
}

export function resolve(fixture: Fixture): VenueWindow {
  assertProgram(fixture, PROGRAM_ADDRESS);
  const supplemental: ReadonlyArray<Address> = fixture.slots
    .slice(BASE_SLOT_COUNT)
    .map((slot) => slot.pubkey);
  assertTailLength(supplemental, 0, 3);
  return build(
    {
      tokenProgramA: slotAt(fixture, 0).pubkey,
      tokenProgramB: slotAt(fixture, 1).pubkey,
      tokenAuthority: slotAt(fixture, 3).pubkey,
      whirlpool: slotAt(fixture, 4).pubkey,
      mintA: slotAt(fixture, 5).pubkey,
      mintB: slotAt(fixture, 6).pubkey,
      tokenOwnerAccountA: slotAt(fixture, 7).pubkey,
      tokenVaultA: slotAt(fixture, 8).pubkey,
      tokenOwnerAccountB: slotAt(fixture, 9).pubkey,
      tokenVaultB: slotAt(fixture, 10).pubkey,
      tickArray0: slotAt(fixture, 11).pubkey,
      tickArray1: slotAt(fixture, 12).pubkey,
      tickArray2: slotAt(fixture, 13).pubkey,
      oracle: slotAt(fixture, 14).pubkey,
    },
    supplemental,
  );
}

export function reachableAccountCounts(): Array<number> {
  const counts: Array<number> = [];
  for (let length = 0; length <= 3; length += 1) {
    const supplemental: ReadonlyArray<Address> = Array.from({ length }, (_, index) =>
      key(20 + index),
    );
    assertTailLength(supplemental, 0, 3);
    counts.push(build(placeholderAccounts(), supplemental).accountCount);
  }
  return counts;
}

File: clients/rust/turk-router/tests/common/harness/whirlpool.rs
use solana_pubkey::Pubkey;
use turk_router::venues::whirlpool::{self, SupplementalTickArrays, WhirlpoolAccounts};
use turk_router::VenueWindow;

use crate::common::fixture::Fixture;

const BASE_SLOT_COUNT: usize = 15;

fn key(byte: u8) -> Pubkey {
    Pubkey::new_from_array([byte; 32])
}

pub fn resolve(fixture: &Fixture) -> VenueWindow {
    assert_eq!(
        fixture.program_id,
        whirlpool::PROGRAM_ID,
        "{}: not a Whirlpool fixture",
        fixture.pool_b58
    );

    let accounts = WhirlpoolAccounts {
        token_program_a: fixture.slot(0).pubkey,
        token_program_b: fixture.slot(1).pubkey,
        token_authority: fixture.slot(3).pubkey,
        whirlpool: fixture.slot(4).pubkey,
        mint_a: fixture.slot(5).pubkey,
        mint_b: fixture.slot(6).pubkey,
        token_owner_account_a: fixture.slot(7).pubkey,
        token_vault_a: fixture.slot(8).pubkey,
        token_owner_account_b: fixture.slot(9).pubkey,
        token_vault_b: fixture.slot(10).pubkey,
        tick_array_0: fixture.slot(11).pubkey,
        tick_array_1: fixture.slot(12).pubkey,
        tick_array_2: fixture.slot(13).pubkey,
        oracle: fixture.slot(14).pubkey,
    };

    let tail: Vec<Pubkey> = fixture
        .slots
        .get(BASE_SLOT_COUNT..)
        .unwrap_or_default()
        .iter()
        .map(|slot| slot.pubkey)
        .collect();
    let supplemental = match tail.as_slice() {
        [] => SupplementalTickArrays::from([]),
        [a] => SupplementalTickArrays::from([*a]),
        [a, b] => SupplementalTickArrays::from([*a, *b]),
        [a, b, c] => SupplementalTickArrays::from([*a, *b, *c]),
        other => panic!(
            "{}: {} supplemental tick arrays, whirlpool accepts at most 3",
            fixture.pool_b58,
            other.len()
        ),
    };

    whirlpool::resolve(&accounts, supplemental)
}

pub fn reachable_account_counts() -> Vec<u8> {
    let accounts = WhirlpoolAccounts {
        token_program_a: key(1),
        token_program_b: key(2),
        token_authority: key(3),
        whirlpool: key(4),
        mint_a: key(5),
        mint_b: key(6),
        token_owner_account_a: key(7),
        token_vault_a: key(8),
        token_owner_account_b: key(9),
        token_vault_b: key(10),
        tick_array_0: key(11),
        tick_array_1: key(12),
        tick_array_2: key(13),
        oracle: key(14),
    };

    [
        SupplementalTickArrays::from([]),
        SupplementalTickArrays::from([key(20)]),
        SupplementalTickArrays::from([key(20), key(21)]),
        SupplementalTickArrays::from([key(20), key(21), key(22)]),
    ]
    .into_iter()
    .map(|supplemental| whirlpool::resolve(&accounts, supplemental).account_count())
    .collect()
}
